File: CICADA_FIGURES/a3a_Fig_Swimmers_RCT.m
clear all; clc; format compact

% FIGURE 1: RCT DATA ONLY (Full-width swimmer plots + RCT survival curves)
figure(1); clf;
T1 = readtable('trialData1.csv');
fcnSingleSwimmerPlot_v4(T1)

%% Get the position of the lowest swimmer plot to align survival curves
all_axes = findall(gcf, 'Type', 'axes');
positions = cell2mat(get(all_axes, 'Position'));
lowest_swimmer_bottom = min(positions(:,2)); % Bottom edge of lowest swimmer plot

% Get the left and width from one of the swimmer plots for alignment
swimmer_left = positions(1,1);  % Left position of swimmer plots
swimmer_width = positions(1,3); % Width of swimmer plots

%% survival curves - computed by a1_CausalSurvivalAnalysis.m
load('bootstrap_confidence_bands.mat');

% Position for bottom plot - use remaining space at bottom
vertical_gap = 0.015; % Small gap between swimmer plots and survival plot
bottom_margin = 0.05; % Small bottom margin
% Calculate height to fill remaining space
height = lowest_swimmer_bottom - vertical_gap - bottom_margin;

% Create axes aligned with swimmer plots above
ax4 = axes('Position', [swimmer_left, bottom_margin, swimmer_width, height]);

% Plot RCT survival curves with SOLID lines
hold on;

% Get reference curves from RCT data
T1_logit = readtable('trialData1.csv');
[s0_true, s1_true, t0_true, t1_true] = fcnPlotKM(T1_logit);

% Time vector
t_grid = 0:2:168;  % This gives 85 points

% Plot main curves - SOLID lines for RCT with blue colors
h_untreated = plot(t0_true, s0_true, '-', 'Color', [0.4, 0.4, 0.8], 'LineWidth', 2.5); % Medium blue
h_treated = plot(t1_true, s1_true, '-', 'Color', [0, 0, 0.5], 'LineWidth', 2.5);    % Dark blue

% Add text labels directly on the curves
% Find good positions for labels (moved further right)
label_time = 130; % Moved even further to the right
[~, idx0] = min(abs(t0_true - label_time));
[~, idx1] = min(abs(t1_true - label_time));

% Add labels with larger offset below the curves
text(label_time, s0_true(idx0) - 0.10, 'Untreated (RCT)', ...
    'Color', [0.4, 0.4, 0.8], 'FontSize', 12, 'FontWeight', 'bold', ...
    'HorizontalAlignment', 'center');
text(label_time, s1_true(idx1) - 0.10, 'Treated (RCT)', ...
    'Color', [0, 0, 0.5], 'FontSize', 12, 'FontWeight', 'bold', ...
    'HorizontalAlignment', 'center');

% Formatting
xlabel('Time (hours)', 'FontSize', 12);
% Move y-axis label inside the plot
text(5, 0.3, 'Survival Probability', 'FontSize', 12, 'FontWeight', 'bold', 'Rotation', 90);
grid on;
set(gca, 'FontSize', 11);
xlim([0, 168]);
ylim([0, 1]);

% Set x-axis ticks every 24 hours
xticks(0:24:168);
xticklabels(0:24:168);
xlim([0 168]);

% Set figure size (in inches)
fig_width = 6;    % Width in inches (adjust as needed)
fig_height = 9;   % Height in inches (adjust as needed)

% Set figure properties
set(gcf, 'Units', 'inches');
set(gcf, 'Position', [1, 1, fig_width, fig_height]); % [left, bottom, width, height]
set(gcf, 'PaperUnits', 'inches');
set(gcf, 'PaperSize', [fig_width, fig_height]);
set(gcf, 'PaperPosition', [0, 0, fig_width, fig_height]);

% Export as PDF (vector format)
print(gcf, 'Fig3_swimmer_survival_plot_RCT.pdf', '-dpdf', '-r300');

%% Export RCT swimmer plot analysis results to text file for paper
filename = sprintf('rct_swimmer_analysis_results_%s.txt', datestr(now, 'yyyymmdd_HHMMSS'));
fid = fopen(filename, 'w');

fprintf(fid, '==========================================================\n');
fprintf(fid, 'RCT SWIMMER PLOT ANALYSIS RESULTS FOR PAPER\n');
fprintf(fid, 'Generated on: %s\n', datestr(now));
fprintf(fid, '==========================================================\n\n');

% Load and analyze RCT data
T1 = readtable('trialData1.csv');
unique_patients = unique(T1.sid);
N = length(unique_patients);

fprintf(fid, 'RCT DATA CHARACTERISTICS:\n');
fprintf(fid, '- Sample size: %d patients\n', N);
fprintf(fid, '- Study type: Randomized Controlled Trial (RCT=1)\n');
fprintf(fid, '- Study period: 168 hours\n');
fprintf(fid, '- Data file: trialData1.csv\n\n');

% Analyze treatment assignment
treated_patients = 0;
untreated_patients = 0;
for i = 1:N
    patient_id = unique_patients(i);
    patient_data = T1(T1.sid == patient_id, :);
    if patient_data.Rx(1) == 1
        treated_patients = treated_patients + 1;
    else
        untreated_patients = untreated_patients + 1;
    end
end

fprintf(fid, 'TREATMENT ASSIGNMENT:\n');
fprintf(fid, '- Treated patients: %d (%.1f%%)\n', treated_patients, treated_patients/N*100);
fprintf(fid, '- Untreated patients: %d (%.1f%%)\n', untreated_patients, untreated_patients/N*100);
fprintf(fid, '- Assignment method: Randomized (50%% probability)\n\n');

% Calculate survival outcomes from Kaplan-Meier analysis
[s0_true, s1_true, t0_true, t1_true] = fcnPlotKM(T1);

% Key time points for analysis
key_times = [24, 48, 72, 96, 120, 144, 168];
fprintf(fid, 'SURVIVAL OUTCOMES AT KEY TIME POINTS:\n');
fprintf(fid, '%-8s %-15s %-15s %-15s\n', 'Time(h)', 'Untreated(%)', 'Treated(%)', 'Difference(%)');
fprintf(fid, '%-8s %-15s %-15s %-15s\n', '-------', '-------------', '-----------', '-------------');

for i = 1:length(key_times)
    time_point = key_times(i);
    
    % Find closest time indices
    [~, idx0] = min(abs(t0_true - time_point));
    [~, idx1] = min(abs(t1_true - time_point));
    
    untreated_surv = s0_true(idx0) * 100;
    treated_surv = s1_true(idx1) * 100;
    difference = treated_surv - untreated_surv;
    
    fprintf(fid, '%-8d %-15.1f %-15.1f %-15.1f\n', ...
        time_point, untreated_surv, treated_surv, difference);
end

% Primary endpoint analysis (168 hours)
final_untreated = s0_true(end) * 100;
final_treated = s1_true(end) * 100;
primary_ate = (s1_true(end) - s0_true(end)) * 100;

fprintf(fid, '\nPRIMARY ENDPOINT (168 hours):\n');
fprintf(fid, '- Untreated survival: %.1f%%\n', final_untreated);
fprintf(fid, '- Treated survival: %.1f%%\n', final_treated);
fprintf(fid, '- Average Treatment Effect (ATE): %.1f%% points\n', primary_ate);

% Clinical interpretation
if primary_ate > 0
    fprintf(fid, '- Treatment effect: BENEFICIAL\n');
    if primary_ate > 10
        fprintf(fid, '- Effect size: LARGE (>10%% points)\n');
    elseif primary_ate > 5
        fprintf(fid, '- Effect size: MODERATE (5-10%% points)\n');
    else
        fprintf(fid, '- Effect size: SMALL (0-5%% points)\n');
    end
else
    fprintf(fid, '- Treatment effect: HARMFUL\n');
end

% Calculate additional survival metrics
fprintf(fid, '\nADDITIONAL SURVIVAL METRICS:\n');

% Calculate deaths in each group
untreated_deaths = 0;
treated_deaths = 0;
total_patient_hours_untreated = 0;
total_patient_hours_treated = 0;

for i = 1:N
    patient_id = unique_patients(i);
    patient_data = T1(T1.sid == patient_id, :);
    patient_data = sortrows(patient_data, 't');
    
    % Determine treatment group
    is_treated = patient_data.Rx(1) == 1;
    
    % Check for death
    death_occurred = any(patient_data.Y > 0);
    
    % Calculate observation time
    max_time = max(patient_data.t);
    
    if is_treated
        if death_occurred
            treated_deaths = treated_deaths + 1;
        end
        total_patient_hours_treated = total_patient_hours_treated + max_time;
    else
        if death_occurred
            untreated_deaths = untreated_deaths + 1;
        end
        total_patient_hours_untreated = total_patient_hours_untreated + max_time;
    end
end

% Calculate mortality rates
untreated_mortality_rate = untreated_deaths / untreated_patients * 100;
treated_mortality_rate = treated_deaths / treated_patients * 100;
relative_risk = treated_mortality_rate / untreated_mortality_rate;

fprintf(fid, '- Untreated deaths: %d/%d (%.1f%%)\n', untreated_deaths, untreated_patients, untreated_mortality_rate);
fprintf(fid, '- Treated deaths: %d/%d (%.1f%%)\n', treated_deaths, treated_patients, treated_mortality_rate);
fprintf(fid, '- Relative risk of death: %.2f\n', relative_risk);

% Number needed to treat/harm
if primary_ate ~= 0
    nnt = 100 / abs(primary_ate);
    if primary_ate > 0
        fprintf(fid, '- Number needed to treat (NNT): %.1f patients\n', nnt);
    else
        fprintf(fid, '- Number needed to harm (NNH): %.1f patients\n', nnt);
    end
end

% Incidence rates (deaths per 1000 patient-hours)
untreated_incidence = (untreated_deaths / total_patient_hours_untreated) * 1000;
treated_incidence = (treated_deaths / total_patient_hours_treated) * 1000;

fprintf(fid, '- Untreated incidence rate: %.2f deaths per 1000 patient-hours\n', untreated_incidence);
fprintf(fid, '- Treated incidence rate: %.2f deaths per 1000 patient-hours\n', treated_incidence);

% Survival curve characteristics
fprintf(fid, '\nSURVIVAL CURVE CHARACTERISTICS:\n');

% Calculate median survival times (if applicable)
median_surv_untreated = NaN;
median_surv_treated = NaN;

% Find median survival (time when survival drops to 50%)
untreated_below_50 = find(s0_true < 0.5, 1);
if ~isempty(untreated_below_50)
    median_surv_untreated = t0_true(untreated_below_50);
    fprintf(fid, '- Median survival (untreated): %.1f hours\n', median_surv_untreated);
else
    fprintf(fid, '- Median survival (untreated): >168 hours (not reached)\n');
end

treated_below_50 = find(s1_true < 0.5, 1);
if ~isempty(treated_below_50)
    median_surv_treated = t1_true(treated_below_50);
    fprintf(fid, '- Median survival (treated): %.1f hours\n', median_surv_treated);
else
    fprintf(fid, '- Median survival (treated): >168 hours (not reached)\n');
end

% Calculate area under the curve (mean survival time over study period)
auc_untreated = trapz(t0_true, s0_true);
auc_treated = trapz(t1_true, s1_true);

fprintf(fid, '- Mean survival time (untreated): %.1f hours\n', auc_untreated);
fprintf(fid, '- Mean survival time (treated): %.1f hours\n', auc_treated);
fprintf(fid, '- Difference in mean survival: %.1f hours\n', auc_treated - auc_untreated);

% RCT validity and design characteristics
fprintf(fid, '\nRCT DESIGN AND VALIDITY:\n');
fprintf(fid, '- Randomization: Balanced (50%% allocation probability)\n');
fprintf(fid, '- Blinding: Not applicable (treatment intervention)\n');
fprintf(fid, '- Follow-up: Complete (168 hours for all patients)\n');
fprintf(fid, '- Primary endpoint: Survival at 168 hours\n');
fprintf(fid, '- Analysis: Kaplan-Meier survival analysis\n');

% Treatment assignment balance check
assignment_balance = abs(treated_patients - untreated_patients);
fprintf(fid, '- Treatment assignment balance: %d patient difference\n', assignment_balance);
if assignment_balance <= N * 0.1  % Within 10% of expected
    fprintf(fid, '  → EXCELLENT randomization balance\n');
elseif assignment_balance <= N * 0.2  % Within 20%
    fprintf(fid, '  → GOOD randomization balance\n');
else
    fprintf(fid, '  → POOR randomization balance - check randomization\n');
end

% Statistical power considerations (post-hoc)
fprintf(fid, '\nSTATISTICAL CONSIDERATIONS:\n');
fprintf(fid, '- Sample size: %d patients (adequate for pilot study)\n', N);
fprintf(fid, '- Effect size observed: %.1f%% points\n', abs(primary_ate));
if abs(primary_ate) > 5
    fprintf(fid, '- Clinical significance: Clinically meaningful\n');
else
    fprintf(fid, '- Clinical significance: Modest effect\n');
end

% Figure generation details
fprintf(fid, '\nFIGURE GENERATION:\n');
fprintf(fid, '- Figure type: Combined swimmer plots and survival curves\n');
fprintf(fid, '- Output file: Fig3_swimmer_survival_plot_RCT.pdf\n');
fprintf(fid, '- Format: PDF vector graphics (300 DPI)\n');
fprintf(fid, '- Dimensions: 6 × 9 inches\n');
fprintf(fid, '- Curve colors: Blue tones (untreated: medium blue, treated: dark blue)\n');
fprintf(fid, '- Data visualization: Individual patient trajectories + survival curves\n');

% Data quality assessment
fprintf(fid, '\nDATA QUALITY ASSESSMENT:\n');

% Check for missing data or anomalies
total_observations = size(T1,1);
patients_with_complete_data = N;  % Assuming all patients have some data

fprintf(fid, '- Total observations: %d\n', total_observations);
fprintf(fid, '- Patients with data: %d/%d (%.1f%%)\n', ...
    patients_with_complete_data, N, patients_with_complete_data/N*100);
fprintf(fid, '- Average observations per patient: %.1f\n', total_observations/N);

% Clinical implications for RCT results
fprintf(fid, '\nCLINICAL IMPLICATIONS:\n');

if primary_ate > 5
    fprintf(fid, '✓ STRONG evidence for treatment benefit\n');
    fprintf(fid, '  - Consider implementation in clinical practice\n');
    fprintf(fid, '  - RCT provides unbiased treatment effect estimate\n');
elseif primary_ate > 0
    fprintf(fid, '⚠️  MODEST evidence for treatment benefit\n');
    fprintf(fid, '  - Consider larger confirmatory trial\n');
    fprintf(fid, '  - Evaluate cost-effectiveness\n');
else
    fprintf(fid, '✗ EVIDENCE of treatment harm\n');
    fprintf(fid, '  - Treatment not recommended\n');
    fprintf(fid, '  - Investigate safety concerns\n');
end

% Research implications
fprintf(fid, '\nRESEARCH IMPLICATIONS:\n');
fprintf(fid, '- RCT provides gold standard comparator for observational studies\n');
fprintf(fid, '- Treatment effect estimate is unbiased by confounding\n');
fprintf(fid, '- Results validate causal inference methods when they agree\n');
fprintf(fid, '- Swimmer plots show individual patient response patterns\n');

% Technical specifications
fprintf(fid, '\nTECHNICAL SPECIFICATIONS:\n');
fprintf(fid, '- Analysis software: MATLAB\n');
fprintf(fid, '- Survival analysis: fcnPlotKM function\n');
fprintf(fid, '- Visualization: fcnSingleSwimmerPlot_v4 function\n');
fprintf(fid, '- Time resolution: 2-hour intervals\n');
fprintf(fid, '- Statistical method: Kaplan-Meier estimator\n');

fclose(fid);
fprintf('RCT swimmer plot analysis results exported to: %s\n', filename);

File: CICADA_FIGURES_PY/a3a_Fig_Swimmers_RCT.py
"""a3a_Fig_Swimmers_RCT.py

Port of CICADA_FIGURES/a3a_Fig_Swimmers_RCT.m.

Produces Fig3_swimmer_survival_plot_RCT.pdf: two swimmer plots (L and A)
plus a Kaplan-Meier survival panel for the RCT (trialData1.csv).
"""

from __future__ import annotations

import os
import sys
import warnings
from datetime import datetime

import matplotlib
matplotlib.use("Agg")

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

HERE = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.abspath(os.path.join(HERE, ".."))
PY_HELPERS = os.path.join(REPO_ROOT, "python")
sys.path.insert(0, PY_HELPERS)

np.random.seed(0)

try:
    from fcnSingleSwimmerPlot_v4 import fcnSingleSwimmerPlot_v4  # type: ignore
    from fcnPlotKM import fcnPlotKM  # type: ignore
except Exception as exc:  # pragma: no cover
    print(f"[a3a] WARNING: could not import helpers: {exc}")
    sys.exit(0)


def _input_csv(name: str) -> str:
    return os.path.join(REPO_ROOT, name)


def main() -> int:
    trial1 = _input_csv("trialData1.csv")
    if not os.path.exists(trial1):
        print(f"[a3a] WARNING: missing input {trial1}; skipping.")
        return 0

    T1 = pd.read_csv(trial1)

    # Build swimmer figure (two panels, L and A).
    fig = fcnSingleSwimmerPlot_v4(T1)
    fig.set_size_inches(6.0, 9.0)

    # Reposition existing swimmer axes into the top 2/3 of the figure and add a
    # survival axis below, matching MATLAB's layout in a3a.
    axes = [a for a in fig.axes if not a.get_label().startswith("<colorbar>")]
    # Filter out colorbars by checking if they have a regular artist pattern.
    swimmer_axes = [a for a in fig.axes if a.get_ylabel() == "" and len(a.images) > 0]

    if len(swimmer_axes) < 2:
        # Fallback: just use first two non-colorbar axes.
        swimmer_axes = [a for a in fig.axes if a not in getattr(fig, "_colorbars", [])][:2]

    # Figure-wide layout
    left_margin = 0.06
    right_margin = 0.10
    top_margin = 0.02
    bottom_margin = 0.05
    vertical_gap = 0.02

    width = 1.0 - left_margin - right_margin
    # Reserve roughly half the figure for the survival plot.
    surv_height = 0.28
    remaining = 1.0 - top_margin - bottom_margin - surv_height - 2 * vertical_gap
    swim_height = remaining / 2.0

    # Re-position the two swimmer axes (top = L, bottom = A) at the top.
    if len(swimmer_axes) >= 2:
        swimmer_axes[0].set_position(
            [left_margin, 1.0 - top_margin - swim_height, width, swim_height]
        )
        swimmer_axes[1].set_position(
            [
                left_margin,
                1.0 - top_margin - 2 * swim_height - vertical_gap,
                width,
                swim_height,
            ]
        )

    # Survival panel
    ax_surv = fig.add_axes(
        [left_margin, bottom_margin, width, surv_height]
    )

    # KM curves from the RCT data
    s0_true, s1_true, t0_true, t1_true = fcnPlotKM(T1)

    ax_surv.plot(
        t0_true, s0_true, "-", color=(0.4, 0.4, 0.8), linewidth=2.5,
        label="Untreated (RCT)",
    )
    ax_surv.plot(
        t1_true, s1_true, "-", color=(0.0, 0.0, 0.5), linewidth=2.5,
        label="Treated (RCT)",
    )

    # On-curve labels (approx; MATLAB places them 0.10 below the curve at t=130)
    label_time = 130.0
    idx0 = int(np.argmin(np.abs(t0_true - label_time)))
    idx1 = int(np.argmin(np.abs(t1_true - label_time)))
    ax_surv.text(
        label_time,
        max(0.0, float(s0_true[idx0]) - 0.10),
        "Untreated (RCT)",
        color=(0.4, 0.4, 0.8),
        fontsize=12,
        fontweight="bold",
        ha="center",
    )
    ax_surv.text(
        label_time,
        max(0.0, float(s1_true[idx1]) - 0.10),
        "Treated (RCT)",
        color=(0.0, 0.0, 0.5),
        fontsize=12,
        fontweight="bold",
        ha="center",
    )

    ax_surv.set_xlabel("Time (hours)", fontsize=12)
    ax_surv.text(
        5.0, 0.3, "Survival Probability",
        fontsize=12, fontweight="bold", rotation=90,
    )
    ax_surv.grid(True)
    ax_surv.tick_params(labelsize=11)
    ax_surv.set_xlim(0, 168)
    ax_surv.set_ylim(0, 1)
    ax_surv.set_xticks(np.arange(0, 169, 24))

    out_pdf = os.path.join(HERE, "Fig3_swimmer_survival_plot_RCT.pdf")
    fig.savefig(out_pdf, format="pdf", dpi=300)
    plt.close(fig)
    print(f"[a3a] Wrote {out_pdf}")

    # ---- text summary ---------------------------------------------------
    stamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    txt = os.path.join(HERE, f"rct_swimmer_analysis_results_{stamp}.txt")

    try:
        unique_patients = np.unique(T1["sid"].values)
        N = int(len(unique_patients))

        treated = 0
        untreated = 0
        treated_deaths = 0
        untreated_deaths = 0
        tot_hours_trt = 0.0
        tot_hours_unt = 0.0
        for pid in unique_patients:
            pdata = T1[T1["sid"] == pid].sort_values("t")
            is_trt = int(pdata["Rx"].iloc[0]) == 1
            died = bool((pdata["Y"] > 0).any())
            mt = float(pdata["t"].max())
            if is_trt:
                treated += 1
                if died:
                    treated_deaths += 1
                tot_hours_trt += mt
            else:
                untreated += 1
                if died:
                    untreated_deaths += 1
                tot_hours_unt += mt

        final_unt = float(s0_true[-1]) * 100.0
        final_trt = float(s1_true[-1]) * 100.0
        ate = final_trt - final_unt

        with open(txt, "w") as f:
            f.write("=" * 58 + "\n")
            f.write("RCT SWIMMER PLOT ANALYSIS RESULTS FOR PAPER\n")
            f.write(f"Generated on: {datetime.now().isoformat(timespec='seconds')}\n")
            f.write("=" * 58 + "\n\n")
            f.write("RCT DATA CHARACTERISTICS:\n")
            f.write(f"- Sample size: {N} patients\n")
            f.write("- Study type: Randomized Controlled Trial (RCT=1)\n")
            f.write("- Study period: 168 hours\n")
            f.write("- Data file: trialData1.csv\n\n")
            f.write("TREATMENT ASSIGNMENT:\n")
            f.write(
                f"- Treated patients: {treated} ({treated/N*100:.1f}%)\n"
            )
            f.write(
                f"- Untreated patients: {untreated} ({untreated/N*100:.1f}%)\n"
            )
            f.write("- Assignment method: Randomized (50% probability)\n\n")
            f.write("PRIMARY ENDPOINT (168 hours):\n")
            f.write(f"- Untreated survival: {final_unt:.1f}%\n")
            f.write(f"- Treated survival: {final_trt:.1f}%\n")
            f.write(f"- Average Treatment Effect (ATE): {ate:.1f}% points\n\n")
            f.write("MORTALITY:\n")
            if untreated > 0:
                f.write(
                    f"- Untreated deaths: {untreated_deaths}/{untreated} "
                    f"({untreated_deaths/untreated*100:.1f}%)\n"
                )
            if treated > 0:
                f.write(
                    f"- Treated deaths: {treated_deaths}/{treated} "
                    f"({treated_deaths/treated*100:.1f}%)\n"
                )
            f.write("\nFIGURE GENERATION:\n")
            f.write("- Output file: Fig3_swimmer_survival_plot_RCT.pdf\n")
            f.write("- Format: PDF vector graphics (300 DPI)\n")
            f.write("- Dimensions: 6 x 9 inches\n")
        print(f"[a3a] Wrote {txt}")
    except Exception as exc:
        warnings.warn(f"[a3a] text summary failed: {exc}")

    return 0


if __name__ == "__main__":
    sys.exit(main())
